space_packet_parser/packets.py

```diff
@@ -369,7 +420,37 @@ def ccsds_generator(
         packet_bytes = read_buffer[current_pos:current_pos + n_bytes_packet]
         current_pos += n_bytes_packet
         # Wrap the bytes in a RawPacketData object that adds convenience methods for parsing the header
-        yield RawPacketData(packet_bytes)
+        ccsds_packet = CCSDSPacketBytes(packet_bytes)
+
+        if not combine_segmented_packets or ccsds_packet.sequence_flags == SequenceFlags.UNSEGMENTED:
+            yield ccsds_packet
+        elif ccsds_packet.sequence_flags == SequenceFlags.FIRST:
+            _segmented_packets[ccsds_packet.apid] = [ccsds_packet]
+            continue
+        elif not _segmented_packets.get(ccsds_packet.apid, []):
+            warnings.warn("Continuation packet found without declaring the start, skipping this packet.")
+            continue
+        elif ccsds_packet.sequence_flags == SequenceFlags.CONTINUATION:
+            _segmented_packets[ccsds_packet.apid].append(ccsds_packet)
+            continue
+        else:  # raw_packet_data.sequence_flags == packets.SequenceFlags.LAST:
+            _segmented_packets[ccsds_packet.apid].append(ccsds_packet)
+            # We have received the final packet, close it up and combine all of
+            # the segmented packets into a single "packet" for XTCE parsing
+            sequence_counts = [p.sequence_count for p in _segmented_packets[ccsds_packet.apid]]
+            if not all((sequence_counts[i + 1] - sequence_counts[i]) % 16384 == 1
+                        for i in range(len(sequence_counts) - 1)):
+                warnings.warn(f"Continuation packets for apid {ccsds_packet.apid} "
+                              f"are not in sequence {sequence_counts}, skipping these packets.")
+                continue
+            # Add all content (including header) from the first packet
+            raw_data = _segmented_packets[ccsds_packet.apid][0]
+            # Add the continuation packets to the first packet, skipping the headers
+            for p in _segmented_packets[ccsds_packet.apid][1:]:
+                raw_data += p[header_length_bytes + secondary_header_bytes:]
+            yield CCSDSPacketBytes(raw_data)
```

Comment:
I think we have a minor memory leak in `_segmented_packets`. We should `pop` the APID key of the `_segmented_packets` dict for this packet in order to yield the combined `CCSDSPacketBytes` object.

Upon further inspection, this probably isn't a huge deal because the biggest this dictionary will ever be is the number of distinct muxed APIDs in the packet stream being parsed. It still feels cleaner to me to pop the continued packet from the dict rather than just leaving it there.

Reply:
Every time we get a new START packet, we create a new storage list. But, I think you're right that this could grow if we get something like LAST, CONTINUATION, CONTINUATION, LAST, CONTINUATION, because there would be packets in the queue still because we never got another START. I added the pop in because I think that is a good way to handle this, thanks for the suggestion!